### ISSUE TITLE
Include Declarations.yaml in Libtorch distributions

### ISSUE THREAD

**Issue by adampauls** [pytorch/pytorch#69471]
## 🚀 Feature
It would be great if the auto-generated Declarations.yaml were available in the pre-built LibTorch distributions. That way, generated external bindings can use the libtorch distributions without additional work. Right now, IIUC, you have to manually run "python -m tools.codegen.gen" after downloading source. 

## Motivation

Make it easier to produce bindings for other languages.

## Pitch

Include Declarations.yaml in Libtorch distributions. 

## Alternatives

The status quo requires download pytorch source and running a quick script, which is not so bad.

**Comment by ezyang** [comment]
This is unlikely to happen, but mostly because we're trying to deprecate Declarations.yaml

**Comment by dfalbel** [comment]
Hi @ezyang, 

The R bindings to LibTorch and AFAICT other bindings (Rust, OCaml, Elixir) depend on Declarations.yaml, is the plan to completely remove it or maybe having another kind of file listing functions in the API?

**Comment by ezyang** [comment]
The plan is to officially publish our codegen as part of PyTorch so it can be used directly! (And in the meantime, Declarations.yaml isn't going anywhere, we're not going to go out of our way to delete it)

**Comment by dfalbel** [comment]
Ohh right! Thanks very much! I was scared that it was going away completely :P

**Comment by adampauls** [comment]
Just a friendly check in on this issue. Is the plan to still publish the output of codegen I don't see it in 1.12.1. Thanks!

**Comment by ezyang** [comment]
I don't remember if we made it for 1.12 but it's definitely coming for 1.13; import torchgen after installing PyTorch

**Comment by adampauls** [comment]
Sorry, what is `import torchgen`? The issue is about publishing the `Declarations.yaml` in the pre-built binary distributions (for Libtorch, which shouldn't require installation of PyTorch?).

**Comment by ezyang** [comment]
It's a library that gives an alternate way of getting at the data that Declarations.yaml produces. It also has a ton of useful support code on top as well

**Comment by adampauls** [comment]
Makes sense. I would stress though that it would be quite nice for libtorch wrappers not to have to install PyTorch (or download source) just to get their hands on Declarations.yaml. 

I suppose I misinterpreted your comment above -- it sounds like you specifically are planning to remove Declarations.yaml, so that if someone wants to autogen bindings for libtorch, they will have to install the matching PyTorch version and then either generate some yaml themselves or use PyTorch codegen as a library, forcing that developer to use Python as the codegen language?

**Comment by ezyang** [comment]
It hasn't been a big problem to keep generating Declarations.yaml, but we don't internally use it for anything, nor do we test that the output makes sense or check that people adding new features represent it properly in the format. So no we won't try to egregiously break you but we also aren't paying attention either.

You can always have a small Python script to dump the real data model to a serial format that your other language bindgen can ingest; we aren't going to tell you what the format should be tho

**Comment by ezyang** [comment]
You are also welcome to keep using Declarations.yaml for as long as it is working for you; the cost being bugs like this are unlikely to get fixed ;)

**Comment by adampauls** [comment]
Right. The specific thing I was trying to avoid was the need to run a small Python script, since that's the status quo. I agree it's better to be able to install PyTorch than download the source, so it's a small improvement, but ah well, it's a minor point.

**Comment by darioteixeira** [comment]
I'll make the case why it's important that the `declarations.yaml` should be included with the source, or **at least** why there should be simple and clear instructions on how to build it.
In my case, I need the `declarations.yaml` file to build Pytorch bindings for a different language. I'm not very familiar with the Python ecosystem, so I just ran the command mentioned at the top of this ticket -- `python -m tools.codegen.gen` -- which immediately failed. I deduce that the source must be compiled beforehand, which is not an easy task if you're not fluent in Python-lore!
